pkg/agent/openflow/pipeline.go

```diff
@@ -2485,38 +2839,109 @@ func (c *client) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAf
+// Tables: L3ForwardingTable
+// Refactored from:
+//   - func (c *client) decTTLFlows(category cookie.Category) []binding.Flow
+// decTTLFlows generates the flow to process TTL. For the packets forwarded across Nodes, TTL should be decremented by one;
+// for packets which enter OVS pipeline from the Antrea gateway, as the host IP stack should have decremented the TTL
+// already for such packets, TTL should not be decremented again.
+func (c *featurePodConnectivity) decTTLFlows(category cookie.Category) []binding.Flow {
 	var flows []binding.Flow
-	for _, proto := range c.ipProtocols {
+	for _, ipProtocol := range c.ipProtocols {
 		flows = append(flows,
 			// Skip packets from the gateway interface.
-			L3DecTTLTable.BuildFlow(priorityHigh).
+			L3DecTTLTable.ofTable.BuildFlow(priorityHigh).
 				Cookie(c.cookieAllocator.Request(category).Raw()).
-				MatchProtocol(proto).
+				MatchProtocol(ipProtocol).
 				MatchRegMark(FromGatewayRegMark).
-				Action().GotoTable(L3DecTTLTable.GetNext()).
+				Action().NextTable().
 				Done(),
-			L3DecTTLTable.BuildFlow(priorityNormal).
+			L3DecTTLTable.ofTable.BuildFlow(priorityNormal).
 				Cookie(c.cookieAllocator.Request(category).Raw()).
-				MatchProtocol(proto).
+				MatchProtocol(ipProtocol).
 				Action().DecTTL().
-				Action().GotoTable(L3DecTTLTable.GetNext()).
+				Action().NextTable().
 				Done(),
 		)
 	}
 	return flows
 }
 
-// externalFlows returns the flows needed to enable SNAT for external traffic.
-func (c *client) externalFlows(nodeIP net.IP, localSubnet net.IPNet, localGatewayMAC net.HardwareAddr, exceptCIDRs []net.IPNet) []binding.Flow {
-	if !c.enableEgress {
-		return nil
+// Feature: Egress
+// Stage: RoutingStage
+// Tables: L3ForwardingTable
+// Stage: PostRoutingStage
+// Tables: SNATTable
+// Refactored from:
+//   - func (c *client) externalFlows(nodeIP net.IP, localSubnet net.IPNet, localGatewayMAC net.HardwareAddr, exceptCIDRs []net.IPNet) []binding.Flow
+// externalFlows generates the flows to perform SNAT for the packets of connections to the external network. The flows identify
+// the packets to external network, and send them to SNATTable, where SNAT IPs are looked up for the packets.
+func (c *featureEgress) externalFlows(category cookie.Category, exceptCIDRs []net.IPNet) []binding.Flow {
+	exceptCIDRsMap := make(map[binding.Protocol][]net.IPNet)
+	for _, cidr := range exceptCIDRs {
+		if cidr.IP.To4() == nil {
+			exceptCIDRsMap[binding.ProtocolIPv6] = append(exceptCIDRsMap[binding.ProtocolIPv6], cidr)
+		} else {
+			exceptCIDRsMap[binding.ProtocolIP] = append(exceptCIDRsMap[binding.ProtocolIP], cidr)
+		}
+	}
+
+	flows := []binding.Flow{
+		// This generates the flow to match the packets to external network and forward them to SNATTable. The packets
+		// from local Pods or remote Pods will hit the flow.
+		// For simplicity, in the following:
+		//   - per-Node IPAM Pod is referred to as Pod.
+		//   - Remote Nodes' IPs are excluded by other flows.
```

Comment:
Actually I didn't find which flow is used to exclude the remote Pods in noEncap mode, as the mark `ToExternalRegMark` is also set in L3Routing flow actions. So could you help point out?

Reply:
See `func (c *featureEgress) snatSkipNodeFlow(category cookie.Category, nodeIP net.IP) binding.Flow`